tests/test_transform.py

```diff
@@ -0,0 +1,250 @@
+import math
+from typing import Any
+
+import numpy
+import pytest
+
+from optuna._transform import _SearchSpaceTransform
+from optuna.distributions import BaseDistribution
+from optuna.distributions import CategoricalDistribution
+from optuna.distributions import DiscreteUniformDistribution
+from optuna.distributions import IntLogUniformDistribution
+from optuna.distributions import IntUniformDistribution
+from optuna.distributions import LogUniformDistribution
+from optuna.distributions import UniformDistribution
+
+
+@pytest.mark.parametrize("transform_log", [True, False])
+@pytest.mark.parametrize("transform_step", [True, False])
+@pytest.mark.parametrize(
+    "param,distribution",
+    [
+        (0, IntUniformDistribution(0, 3)),
+        (1, IntLogUniformDistribution(1, 10)),
+        (2, IntUniformDistribution(0, 10, step=2)),
+        (0.0, UniformDistribution(0, 3)),
+        (1.0, LogUniformDistribution(1, 10)),
+        (0.2, DiscreteUniformDistribution(0, 1, q=0.2)),
+        ("foo", CategoricalDistribution(["foo"])),
+        ("bar", CategoricalDistribution(["foo", "bar", "baz"])),
+    ],
+)
+def test_search_space_transform_shapes_dtypes(
+    transform_log: bool,
+    transform_step: bool,
+    param: Any,
+    distribution: BaseDistribution,
+) -> None:
+    trans = _SearchSpaceTransform({"x0": distribution}, transform_log, transform_step)
+    trans_params = trans.transform({"x0": param})
+
+    if isinstance(distribution, CategoricalDistribution):
+        expected_bounds_shape = (len(distribution.choices), 2)
+        expected_params_shape = (len(distribution.choices),)
+    else:
+        expected_bounds_shape = (1, 2)
+        expected_params_shape = (1,)
+    assert trans.bounds.shape == expected_bounds_shape
+    assert trans.bounds.dtype == numpy.float64
+    assert trans_params.shape == expected_params_shape
+    assert trans_params.dtype == numpy.float64
+
+
+@pytest.mark.parametrize("transform_log", [True, False])
+@pytest.mark.parametrize("transform_step", [True, False])
```

Comment:
How about separating a test method that checks these parameters.
It makes the test cases simple even though it increases the test code lines.

Reply:
I couldn't understand this comment but do you mean to write two separate test functions, one for `transform_log` and one for `transform_step`?

Reply:
Yes. This is not a strong opinion.
It makes the tests simple and error messages will be clear.

Reply:
I actually think we should test them and their permutations together. 

On a side note, we could probably omit this parameterizations from the other test functions. That might make the entire test file more comprehensible. What do you think?

Reply:
Pushed a change with that, PTAL.

Reply:
Thank you for the update. It looks current test case is enough simple 👍  Good refactoring.